### ISSUE TITLE
Feature Request: Point in Time Recovery (PITR)

### ISSUE THREAD

**Issue by bdurrow** [sorintlab/stolon#144]
I am interested in bringing up production postgres databases in OpenShift Origin.  It looks like there has been some work to template stolon for OpenShift so it might be a good fit.  Part of our requirements include Point In Time Recovery (PITR) but I don't see any backup/restore features in stolon.  Is it on the roadmap?

I have done some work with pghoard (https://github.com/ohmu/pghoard) to have it work in a docker and plan to do more work to allow it to serve archived WALs for recovery out of a docker (with object store backing).  Does this fit architecturally with Stolon?

**Comment by wied03** [comment]
It doesn't fix this, but I had a rough procedure I used to use with corosync/pacemaker for this that worked. I used WAL-E for backup so this kind of depended on that. Maybe this might help the process though.

The rough steps were:
1. Shutdown the corosync cluster
2. Wipe out the PGSQL data directory
3. Setup a recovery.conf as noted below
4. Run wal-e backup-fetch
5. Start just a single node (the master) without the corosync/pacemaker stuff (the standard pgsql way)
6. Verify the restore looks good on the master
7. Run ```select pg_xlog_replay_resume();``` on the master
8. Wait for ```select pg_is_in_recovery();`` to indicate recovery is over
9. Once recovery is over, shutdown the standalone instance and start the master via Pacemaker (but with slaves turned off)
10. Once the master comes up OK, Run pg_basebackup on the slaves to get a complete baseline set of data again
11. Cleanup the Pacemaker resource state
12. Re-enable the slaves

Recovery.conf:
```bash
 restore_command='envdir /etc/wal-e.d/env /usr/bin/wal-e wal-fetch %f %p'
 recovery_target_name = '<%=restore_point_name%>'
 # If we don't include latest, the default is the timeline of the base backup, which in a re-restore situation, won't work since we will have created a new timeline by then
 recovery_target_timeline = 'latest'
 ```

**Comment by sgotti** [comment]
@bdurrow sorry for the late reply.

> I am interested in bringing up production postgres databases in OpenShift Origin. It looks like there has been some work to template stolon for OpenShift so it might be a good fit. Part of our requirements include Point In Time Recovery (PITR) but I don't see any backup/restore features in stolon. Is it on the roadmap?

There's some work going on backup feature.

About the restore feature I'd like to know some details on your use case.

Are you interested on restoring a db on a running stolon cluster or just restoring at cluster initialization?

**Comment by bdurrow** [comment]
@sgotti,

We have developed an EJB/Hibernate application that persists to postgres.  We develop and run this software in an openshift cluster and are looking for a way to recover from (mostly human caused) disaster.  If we loose the Stolon cluster (or the entire openshift cluster) we also need a way to recover to the latest available version of the data.

Use cases (both would hopefully end up with a working Stolon cluster):

1) Bug or engineer error causes data loss: we need PITR recovery to just before things went south.  We would also need to recover any data that was captured after the recovery point but would expect to do that manually.

2) Total loss of cluster: Initialize a new cluster and recover most recent backup.  Hopefully with a very low Recovery Point Objective (less than 2 minutes).

3) Build verification: Make a new Stolon cluster with the most recent data for use in a staging environment to validate a build before we push to production.

**Comment by sgotti** [comment]
@bdurrow @wied03 Thanks for the inputs.

Currently you can do a db recovery but this needs to manually stop stolon keepers and sentinels, do recovery and then restart the processes (in the right order). This can be improved adding the ability to "freeze" the cluster, do manual operations and then "unfreeze" it (I'll open an issue for this).

I think that stolon could be improved to orchestrate a recovery without manual intervention. But before introducing a stolon backup/restore solution I'll prefer to easily integrate with other backup tools (like wal-e).
An idea that I'm thinking about for the PITR part will be something like this:

* Define in the cluster config:
 * A backup restore script (with some parametrizable arguments like the datadir etc...)
 * The wal restore_command (for the recovery.conf)
* A new cluster state called "recovery" (activated via stolonctl passing optional recovery target and other arguments for the restore script) that will basically do the recovery on a keeper (calling the backup restore script and configuring recovery.conf) and make the other slaves resync on the restored db.

Thoughts?

**Comment by wied03** [comment]
My entire PITR process i wrote about was based around wal-e so I agree
there. It seems like pacemaker has the same idea of a recovery state,
probably for a good reason.

sent from my mobile device

**Comment by sgotti** [comment]
In the last weeks a lot of changes have been implemented to make stolon able to provide a lot of new features. These will land in v0.5.0. Since a lot of things and also the UX has changed I'll be happy if you can test them and provide any feedback (please open new issues/PR). 

One of these changes is support for PITR (#183). The main goal is to support any kind of backup/restore tool. The related doc is [here](https://github.com/sorintlab/stolon/blob/master/doc/pitr.md) and there's also a [wal-e example](https://github.com/sorintlab/stolon/blob/master/doc/pitr_wal-e.md).

**Comment by sgotti** [comment]
I'm going to close this since it was implemented in #183 and also a wal-e example is provided. Please open new issues/PRs if you encounter some issues or want to propose some improvements.
